pipeline_tools/confirm_submission.py

```diff
@@ -2,44 +2,66 @@
 
 import requests
 import argparse
-import time
-
-def run(envelope_url, retry_seconds, timeout_seconds):
-    start = time.time()
-    while True:
-        print('Getting status for {}'.format(envelope_url))
-        envelope_js = get_envelope_json(envelope_url)
-        status = envelope_js.get('submissionState')
-        print('submissionState: {}'.format(status))
-        if status == 'Valid':
-            break
-        time.sleep(retry_seconds)
-        current = time.time()
-        if current - start >= timeout_seconds:
-            message = 'Timed out while waiting for Valid status. Timeout seconds: {}'.format(timeout_seconds)
-            raise ValueError(message)
-    confirm(envelope_url)
+from tenacity import retry, retry_if_result, retry_if_exception_type, wait_exponential, stop_after_delay, RetryError
 
+RETRY_SECONDS = 10
+TIMEOUT_SECONDS = 600
+
+
+def status_is_invalid(status):
+    return status != 'Valid'
+
+
+@retry(reraise=True, wait=wait_exponential(multiplier=1, max=RETRY_SECONDS), stop=stop_after_delay(TIMEOUT_SECONDS),
+       retry=(retry_if_result(status_is_invalid) | retry_if_exception_type(requests.HTTPError)))
+def wait_for_valid_status(envelope_url):
+    """
+    Check the status of the submission. Retry until the status is "Valid", or if there is an error with the request to
+    get the submission envelope.
+    :param str envelope_url: Submission envelope url
+    :return: str status: Status of the submission ("Valid", "Validating", etc.)
+    """
+    print('Getting status for {}'.format(envelope_url))
+    envelope_js = get_envelope_json(envelope_url)
+    status = envelope_js.get('submissionState')
+    print('submissionState: {}'.format(status))
+    return status
+
+
+@retry(reraise=True, wait=wait_exponential(multiplier=1, max=RETRY_SECONDS), stop=stop_after_delay(TIMEOUT_SECONDS))
 def confirm(envelope_url):
     print('Confirming submission')
     headers = {
         'Content-type': 'application/json'
     }
     response = requests.put('{}/submissionEvent'.format(envelope_url), headers=headers)
+    response.raise_for_status()
```

Comment:
dumb question: would the errors threw here break the retry?

Reply:
Not a dumb question at all! Throwing an error here will cause the function to retry until there is a successful response and timeout after the specified `TIMEOUT_SECONDS`. If it times out, it will raise the HTTP error from the request.